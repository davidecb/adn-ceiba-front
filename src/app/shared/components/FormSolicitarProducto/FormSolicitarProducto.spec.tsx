import * as React from 'react';
import { RenderResult, fireEvent, render, wait } from '@testing-library/react';
import { SinonStub, stub } from 'sinon';
import { FormSolicitarProducto } from './';

describe('FormSolicitarProducto test', () => {
  let componentWrapper: RenderResult;
  let componentProps: React.ComponentProps<typeof FormSolicitarProducto> & {
    onSubmit: SinonStub;
  };

  beforeEach(() => {
    componentProps = {
      producto: {},
      onSubmit: stub(),
      cerrarFormulario: stub()
    };
    componentWrapper = render(<FormSolicitarProducto {...componentProps} />);
  });

  it('should match snapshot', () => {
    expect(componentWrapper.container).toMatchSnapshot();
  });

  it('should fail on submit all fields missing', async () => {
    const elem = componentWrapper.container;
    const submitButton = elem.querySelector('button[type="submit"]');
    
    await wait(() => {
      submitButton && fireEvent.click(submitButton);
    });
    const spans = elem.querySelectorAll('span');
    expect(spans.length).toBe(3);
    expect(spans[1].textContent).toBe('El campo material es requerido.');
    expect(spans[2].textContent).toBe('El campo color es requerido.');
  });

  it('should fail on submit one field missing', async () => {
    const elem = componentWrapper.container;

    const material = elem.querySelector('select[name="material"]');
    const submitButton = elem.querySelector('button[type="submit"]');

    await wait(() => {
      material && fireEvent.change(material, { target: { value: 'PLA'}});
    });

    await wait(() => {
      submitButton && fireEvent.click(submitButton);
    });
    const spans = elem.querySelectorAll('span');
    expect(spans.length).toBe(2);
    expect(spans[1].textContent).toBe('El campo color es requerido.');
  });

  it('should submit', async () => {
    const elem = componentWrapper.container;

    const material = elem.querySelector('select[name="material"]');
    const color = elem.querySelector('select[name="color"]');
    const submitButton = elem.querySelector('button[type="submit"]');

    await wait(() => {
      material && fireEvent.change(material, { target: { value: 'ABS'}});
    });
    await wait(() => {
      color && fireEvent.change(color, { target: { value: 'madera'}});
    });

    await wait(() => {
      submitButton && fireEvent.click(submitButton);
    });

    const formSubmitted = componentProps.onSubmit.firstCall.args[0];

    expect(formSubmitted.material).toBe('ABS');
    expect(formSubmitted.color).toBe('madera');
    expect(formSubmitted.cantidad).toBe(1);
    expect(formSubmitted.acabado.pulido).toBe(false);
    expect(formSubmitted.acabado.pintado).toBe(false);
    expect(formSubmitted.acabado.barnizado).toBe(false);
    expect(formSubmitted.urgencia).toBe(false);
  });
});
