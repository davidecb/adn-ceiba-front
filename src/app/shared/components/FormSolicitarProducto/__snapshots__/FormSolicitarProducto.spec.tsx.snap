// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSolicitarProducto test should match snapshot 1`] = `
<div>
  <div
    class="sc-AxjAm dUnTGy"
  >
    <span
      class="sc-AxirZ gMixma"
    >
      ❌
    </span>
    <form
      class="sc-AxhCb eFvPgw"
    >
      <h2>
        Personalizar 
      </h2>
      <select
        class="sc-Axmtr fVKlMS"
        name="material"
        placeholder="material"
      >
        <option
          value=""
        >
          Material
        </option>
        <option
          value="PLA"
        >
          PLA
        </option>
        <option
          value="ABS"
        >
          ABS
        </option>
      </select>
      <select
        class="sc-Axmtr fVKlMS"
        name="color"
        placeholder="color"
      >
        <option
          value=""
        >
          Color
        </option>
        <option
          value="negro"
        >
          Negro
        </option>
        <option
          value="blanco"
        >
          Blanco
        </option>
        <option
          value="negro mate"
        >
          Negro mate
        </option>
        <option
          value="madera"
        >
          Madera
        </option>
        <option
          value="plata"
        >
          Plata
        </option>
      </select>
      <div
        class="sc-AxhUy colrEl"
      >
        Acabados:
        <label
          class="sc-AxiKw jXpQYs"
        >
          Pulido 
          <input
            name="acabado"
            type="checkbox"
            value="pulido"
          />
        </label>
        <label
          class="sc-AxiKw jXpQYs"
        >
          Pintado 
          <input
            name="acabado"
            type="checkbox"
            value="pintado"
          />
        </label>
        <label
          class="sc-AxiKw jXpQYs"
        >
          Barnizado 
          <input
            name="acabado"
            type="checkbox"
            value="barnizado"
          />
        </label>
      </div>
      <label
        class="sc-AxiKw jXpQYs"
      >
        Urgencia: 
        <input
          name="acabado"
          type="checkbox"
          value="urgencia"
        />
      </label>
      <div
        class="sc-AxmLO VxQKP"
      >
        <div
          class="sc-fzozJi kzamqH"
        >
          ➖
        </div>
        <div
          class="sc-fzoLsD fzRNOu"
        >
          1
        </div>
        <div
          class="sc-fzozJi kzamqH"
        >
          ➕
        </div>
      </div>
      <button
        class="sc-AxheI ePbRpe"
        type="submit"
      >
        Agregar al carrito
      </button>
    </form>
  </div>
</div>
`;
